control/pid.go

```diff
@@ -50,41 +53,102 @@ func (p *basicPID) Next(ctx context.Context, x []*Signal, dt time.Duration) ([]*
 	p.mu.Lock()
 	defer p.mu.Unlock()
 	if p.tuning {
-		out, done := p.tuner.pidTunerStep(math.Abs(x[0].GetSignalValueAt(0)), p.logger)
-		if done {
-			p.kD = p.tuner.kD
-			p.kI = p.tuner.kI
-			p.kP = p.tuner.kP
-			p.logger.Info("\n\n-------- ***** PID GAINS CALCULATED **** --------")
-			p.logger.CInfof(ctx, "Calculated gains are p: %1.6f, i: %1.6f, d: %1.6f", p.kP, p.kI, p.kD)
-			p.logger.CInfof(ctx, "You must MANUALLY ADD p, i and d gains to the robot config to use the values after tuning\n\n")
-			p.tuning = false
+		// Multi Input/Output Implementation
+		if p.useMulti {
+
+			// For each PID Set and its respective Tuner Object, Step through an iteration of tuning until done.
+			for i := 0; i < len(p.PIDSets); i++ {
```

Comment:
just as an fyi I don't think tuning simultaneously would work for the base. or if you wanted to tune motors and a base? although maybe this change will allow it to do so

Reply:
i kinda agree with you. I think I might try changing this to tune signals one at a time

Reply:
updated the code to tune one signal at a time. the tests are alittle weird with this change(have to turn off tuning after we reach the "end" of the test) but otherwise i think this should work the way we want